File: Problem2.java
public static Map<String, List<String>> favoritegenre(Map<String, List<String>> userMap, Map<String, List<String>> genreMap) {
    Map<String, List<String>> result = new HashMap<>();
    HashMap<String, String> songsToGenre = new HashMap<>();
    for(String genre : genreMap.keySet()){
        List<String> songs = genreMap.get(genre);
        for(String sond : songs){
            songsToGenre.put(song,genre);
        }
    }
    for(String user: userMap.keySet()){
        result.put(user,new ArrayList<>());
        HashMap<String, Integer> countMap = new HashMap<>();
        List<String> songs = userMap.get(user);
        int max = 0;
        for(String song : songs){
            String genre = songsToGenre.get(song);
            countMap.put(genre, countMap.getOrDefault(genre,0) + 1);
            max = Math.max(max, countMap.get(genre));
        }
        for(String genre : countMap.keySet()){
            if(countMap.get(genre) == max){
                result.get(user).add(genre);
            }
        }
    }
    return result;
}

public static void main(String[] args) {
        HashMap<String, List<String>> userSongs = new HashMap<>();

        userSongs.put("David", Arrays.asList(new String[]{"song1", "song2", "song3", "song4", "song8"}));

        userSongs.put("Emma", Arrays.asList(new String[]{"song5", "song6", "song7"}));

        HashMap<String, List<String>> songGenres = new HashMap<>();

        songGenres.put("Rock", Arrays.asList(new String[]{"song1", "song3"}));

        songGenres.put("Dubstep", Arrays.asList(new String[]{"song7"}));

        songGenres.put("Techno", Arrays.asList(new String[]{"song2", "song4"}));

        songGenres.put("Pop", Arrays.asList(new String[]{"song5", "song6"}));

        songGenres.put("Jazz", Arrays.asList(new String[]{"song8", "song9"}));

        Map<String, List<String>> res = favoritegenre(userSongs, songGenres);

        System.out.println(res);
}

File: Problem2.py
def fav_genre(user_map, genre_map):
    song_genre_map = {}
    for key, value in genre_map.items():
        for song in value:
            song_genre_map[song] = key
    ans = {}
    for key, value in user_map.items():
        temp = {}
        user_ans = []
        mx = 0
        for song in value:
            if song_genre_map[song] not in temp:
                temp[song_genre_map[song]] = 1
            else:
                temp[song_genre_map[song]] += 1 
            if temp[song_genre_map[song]] > mx:
                user_ans = [song_genre_map[song]]
                mx = temp[song_genre_map[song]]
            elif temp[song_genre_map[song]] == mx:
                user_ans.append(song_genre_map[song])
        ans[key] = user_ans
    return ans
        
        
        
